execution/src/execution.cc

```diff
@@ -261,15 +277,30 @@ namespace leatherman { namespace execution {
 
         out_stream.open(out_file.c_str(), std::ios::binary);
         if (!out_stream.is_open()) {
-            throw execution_exception(_("failed to open the output file."));
+            throw execution_exception(_("failed to open output file {1}", out_file));
+        }
+
+        boost::system::error_code ec;
+        if (perms) {
+            fs::permissions(out_file, *perms, ec);
+            if (ec) {
+                throw execution_exception(_("failed to modify permissions on output file {1} to {2,num,oct}: {3}", out_file, *perms, ec.message()));
+            }
         }
 
         if (err_file.empty()) {
             setup_execute(stderr_callback, actual_options);
         } else {
             err_stream.open(err_file.c_str(), std::ios::binary);
             if (!err_stream.is_open()) {
-                throw execution_exception(_("failed to open the error file."));
+                throw execution_exception(_("failed to open error file {1}", err_file));
+            }
+
+            if (perms) {
+                fs::permissions(err_file, *perms, ec);
```

Comment:
I wish there was a way in C++ to open file and set its permissions in a single step (like in C's [`open`](http://man7.org/linux/man-pages/man2/open.2.html) function). As if done in two steps like here, there is still a short window when someone can open the file before its permissions are set more restrictively.